### ISSUE TITLE
PhysicsDirectBodyState is singleton?

### ISSUE THREAD

**Issue by steelywing** [godotengine/godot#56245]
### Godot version

3.4

### System information

Windows 10

### Issue description

I get 2 `PhysicsDirectBodyState` of 2 `PhysicsBody` using `PhysicsServer.BodyGetDirectState()`, but they became the same object.

The doc doesn't mention this thing.

### Steps to reproduce

```cs
    public override void _PhysicsProcess(float delta)
    {
        var state = PhysicsServer.BodyGetDirectState(this.chest.GetRid());
        var state2 = PhysicsServer.BodyGetDirectState(this.hips.GetRid());
        GD.Print($"state {state == state2}"); // true, they are same object
    }
```

### Minimal reproduction project

_No response
[PhysicsDirectBodyState.zip](https://github.com/godotengine/godot/files/7776307/PhysicsDirectBodyState.zip)
_

**Comment by Calinou** [comment]
@steelywing Can you reproduce this in GDScript instead of C#?

**Comment by steelywing** [comment]
> @steelywing Can you reproduce this in GDScript instead of C#?

Yes

```
extends Spatial


# Declare member variables here. Examples:
# var a = 2
# var b = "text"


# Called when the node enters the scene tree for the first time.
func _ready():
	pass # Replace with function body.


# Called every frame. 'delta' is the elapsed time since the previous frame.
#func _process(delta):
#	pass
func _physics_process(delta):
	var r1 = PhysicsServer.body_get_direct_state($RigidBody.get_rid())
	var r2 = PhysicsServer.body_get_direct_state($RigidBody2.get_rid())
	r1.linear_velocity += Vector3.FORWARD
	r2.linear_velocity += Vector3.FORWARD
```

Only `r2` move, `r1` doesn't move. I have create a demo in the first post.

**Comment by Zylann** [comment]
That's indeed how this was designed...
```cpp
	static BulletPhysicsDirectBodyState3D *get_singleton(RigidBodyBullet *p_body) {
		singleton->body = p_body;
		return singleton;
	}
```
Allowing more than one to exist requires either creating new instances of the object everytime, or have a sort of pool, or change the API to take the body as parameter of each function.

**Comment by steelywing** [comment]
How do I work with 2 `PhysicsDirectBodyState` at the same time?

I cannot access `BulletPhysicsDirectBodyState` class in C#, so I cannot instance it. I only see abstract class `PhysicsDirectBodyState`.

**Comment by Zylann** [comment]
I know, it's an implementation detail. You get the abstract class in your script but it is in reality one of the two specializations depending on which physics engine you use.

If you want to use Bullet, the only workaround I see is to update the bodies one at a time, and not have both at the same time. Given your snippet, you could rewrite it like this:
```gdscript
	var r1 = PhysicsServer.body_get_direct_state($RigidBody.get_rid())
	r1.linear_velocity += Vector3.FORWARD
	# Done with r1

	var r2 = PhysicsServer.body_get_direct_state($RigidBody2.get_rid())
	r2.linear_velocity += Vector3.FORWARD
	# done with r2
```

Otherwise this needs to be fixed, or you have to use GodotPhysics.

**Comment by madmiraal** [comment]
Duplicate of #42877. This is fixed in 3.x (i.e. in 3.5) with #42929. There was a master branch version (#42928) which included the fix for Bullet too, but it was closed because #49471 was merged instead.
